### PR TITLE
Provide a specialized interval tree for resource management

### PR DESCRIPTION
**Pull request by jiangliu**
This PR implements a specialized version of interval tree for resource management, which could be used manage normal VMM resources.

### CODE REVIEW

File: src/interval_tree.rs [hunk 1]
```diff
@@ -349,6 +442,32 @@ impl<T> IntervalTree<T> {
...
+    /// ```rust
+    /// extern crate vm_allocator;
+    /// use vm_allocator::{IntervalTree, Range, Constraint};
+    ///
+    /// let mut tree = IntervalTree::<u64>::new();
+    /// tree.insert(Range::new(0x100, 0x100), None);
+    /// tree.insert(Range::new(0x200, 0x2ff), None);
+    ///
+    /// let constraint = Constraint::new(2);
+    /// let key = tree.allocate(&constraint);
```
Comment (liujing2) [inline src/interval_tree.rs]: nit: Seems no Constraint definition in the first and second commit so far :)

File: src/interval_tree.rs [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+    }
+}
+
+/// Internal tree node to implement interval tree.
+#[derive(Debug)]
+struct InnerNode<T> {
+    /// Interval handled by this node.
+    key: Range,
+    /// Optional contained data, None if the node is free.
+    data: Option<T>,
```
Comment (liujing2) [inline src/interval_tree.rs]: What is `data` designed for? I can assume the usage, for example, IO address space [start, end] is described by `key` as [start, end], ioapic irq number N is described by `key` as [N, N].

Comment (jiangliu) [reply src/interval_tree.rs]: It could be used to hold memfd for guest memory.

File: src/lib.rs [hunk 3]
```diff
@@ -1,6 +1,43 @@
...
+    pub max: u64,
+    /// Alignment for the allocated resource.
+    pub align: u64,
+    /// Resource allocation policy.
+    pub policy: AllocPolicy,
+}
+
+impl Constraint {
+    /// Create a new constraint object with default settings.
+    pub fn new(size: u64) -> Self {
```
Comment (liujing2) [inline src/lib.rs]: It seems this function needs more parameters?

File: src/interval_tree.rs (hunk 2 above)
Comment (liujing2) [reply src/interval_tree.rs]: Got it. Thanks for explanation.

File: src/lib.rs (hunk 3 above)
Comment (jiangliu) [reply src/lib.rs]: It's used to cover the most common usage scenario. For rare use cases, you may dirrectly construct a Constraint object or we may add some new interfaces for those cases.

Comment (liujing2) [reply src/lib.rs]: So what is the difference between this and `ResourceConstraint` defined in https://github.com/rust-vmm/vm-device/pull/9 ?

Comment (jiangliu) [reply src/lib.rs]: It's two related abstractions. With Device::ResourceConstraint and vm_allocator::Constraint, there's still one piece missing of the whole puzzle. We need the vmm to implement a resource manager, which converts Device::ResourceConstraint to vm_allocator::Constraint.
1) vm_allocator is just a generic resource allocator,
2) vmm::ResourceManager needs to define type of resources and use vm_allocator to manage its resources.
It's still unclear whether vmm::ResourceManager should be VMM specific or we need another vmm_resource crate.

Comment (liujing2) [reply src/lib.rs]: > It's two related abstractions. With Device::ResourceConstraint and vm_allocator::Constraint, there's still one piece missing of the whole puzzle. We need the vmm to implement a resource manager, which converts Device::ResourceConstraint to vm_allocator::Constraint.
> 
Ok, this is the "glue" part between vm-device and vm-allocator.

> 1. vm_allocator is just a generic resource allocator,
> 2. vmm::ResourceManager needs to define type of resources and use vm_allocator to manage its resources.
>    It's still unclear whether vmm::ResourceManager should be VMM specific or we need another vmm_resource crate.

I think we also need a conversion from the vmm::ResourceManager::allocate_resources() return type to vm_device::Resource for registering devices.

File: src/interval_tree.rs (hunk 2 above)
Comment (liujing2) [reply src/interval_tree.rs]: For other resources that don't need this `data`, what is the specific type of T?

Comment (jiangliu) [reply src/interval_tree.rs]: A simple unit struct "()" should be enough.

File: src/lib.rs (hunk 3 above)
Comment (jiangliu) [reply src/lib.rs]: Yes, that's true. Interval tree just manages a set of IDs, the resource manager needs to define the specific resources the ID stands for.

**Comment by sameo** [comment]
Before looking deeper into the actual code, I guess my first general question would be: why couldn't we use one of the existing rust collection structure as the inner `IntervalTree` tree structure:

```rust
struct IntervalTree<T> {
          root: BTreeMap<Range, T>>,          
}
```

**Comment by jiangliu** [comment]
> Before looking deeper into the actual code, I guess my first general question would be: why couldn't we use one of the existing rust collection structure as the inner `IntervalTree` tree structure:
> 
> ```rust
> struct IntervalTree<T> {
>           root: BTreeMap<Range, T>>,          
> }
> ```

IntervalTree supports using `range(start, end)` as the key, so we could search the entry containing a specific value(point).

**Comment by AlexandruCihodaru** [comment]
Closing this as it is obsolete.

**Pull request closed without merging**